crates/network/src/ethereum/mod.rs

```diff
@@ -0,0 +1,267 @@
+    }
+
+    fn set_to(&mut self, to: alloy_primitives::TxKind) {
+        match to {
+            TxKind::Create => self.to = None,
+            TxKind::Call(to) => self.to = Some(to),
+        }
+    }
+
+    fn value(&self) -> Option<alloy_primitives::U256> {
+        self.value
+    }
+
+    fn set_value(&mut self, value: alloy_primitives::U256) {
+        self.value = Some(value)
+    }
+
+    fn gas_price(&self) -> Option<U256> {
+        self.gas_price
+    }
+
+    fn set_gas_price(&mut self, gas_price: U256) {
+        self.gas_price = Some(gas_price);
+    }
+
+    fn max_fee_per_gas(&self) -> Option<U256> {
+        self.max_fee_per_gas
+    }
+
+    fn set_max_fee_per_gas(&mut self, max_fee_per_gas: U256) {
+        self.max_fee_per_gas = Some(max_fee_per_gas);
+    }
+
+    fn max_priority_fee_per_gas(&self) -> Option<U256> {
+        self.max_priority_fee_per_gas
+    }
+
+    fn set_max_priority_fee_per_gas(&mut self, max_priority_fee_per_gas: U256) {
+        self.max_priority_fee_per_gas = Some(max_priority_fee_per_gas);
+    }
+
+    fn max_fee_per_blob_gas(&self) -> Option<U256> {
+        self.max_fee_per_blob_gas
+    }
+
+    fn set_max_fee_per_blob_gas(&mut self, max_fee_per_blob_gas: U256) {
+        self.max_fee_per_blob_gas = Some(max_fee_per_blob_gas)
+    }
+
+    fn gas_limit(&self) -> Option<U256> {
+        self.gas
+    }
+
+    fn set_gas_limit(&mut self, gas_limit: U256) {
+        self.gas = Some(gas_limit);
+    }
+
+    fn build_unsigned(self) -> BuilderResult<<Ethereum as Network>::UnsignedTx> {
+        match (
+            self.gas_price.as_ref(),
+            self.max_fee_per_gas.as_ref(),
+            self.access_list.as_ref(),
+            self.max_fee_per_blob_gas.as_ref(),
+            self.blob_versioned_hashes.as_ref(),
+            self.sidecar.as_ref(),
+        ) {
+            // Legacy transaction
+            (Some(_), None, None, None, None, None) => build_legacy(self).map(Into::into),
+            // EIP-2930
+            // If only accesslist is set, and there are no EIP-1559 fees
+            (_, None, Some(_), None, None, None) => build_2930(self).map(Into::into),
+            // EIP-1559
+            // If EIP-4844 fields are missing
+            (None, _, _, None, None, None) => build_1559(self).map(Into::into),
+            // EIP-4844
+            // All blob fields required
+            (None, _, _, Some(_), Some(_), Some(_)) => {
+                build_4844(self).map(TxEip4844Variant::from).map(Into::into)
+            }
+            _ => build_legacy(self).map(Into::into),
+        }
+    }
+
+    async fn build<S: NetworkSigner<Ethereum>>(
+        self,
+        signer: &S,
+    ) -> BuilderResult<<Ethereum as Network>::TxEnvelope> {
+        // todo: BuilderResult + SignerResult
+        Ok(signer.sign(self.build_unsigned()?).await.unwrap())
+    }
+}
+
+/// Build a legacy transaction.
+fn build_legacy(request: TransactionRequest) -> Result<TxLegacy, TransactionBuilderError> {
```

Comment:
Ideally I'd have these on `TransactionRequest`, but that would cause a circular dep (`alloy-network` depends on `rpc-types` -> `rpc-types` would need to depend on `alloy-network` for `TransactionBuilderError`)